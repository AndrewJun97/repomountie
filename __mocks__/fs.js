// __mocks__/fs.js
'use strict';

const fs = jest.requireActual('fs');

function access(path, flag, cb) {
  cb(undefined);
}

function readFile(path, options, cb) {
  cb(undefined, Buffer.from('Hello World', 'utf8'));
  // cb(undefined, 'Hello World');
}

fs.readFile = readFile;
fs.access = access;

module.exports = fs;
