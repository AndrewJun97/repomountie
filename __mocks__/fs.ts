// __mocks__/fs.js
'use strict';

const fs = require.requireActual('fs');

function access(path, flag, cb) {
  if (path === 'no-file-access') {
    return cb(new Error('No access to this file - mock'));
  }

  return cb(undefined);
}

function readFile(path, options, cb) {
  if (path === 'no-file') {
    return cb(new Error('No such file - mock'), undefined);
  }
  const text = 'Hello World';
  return cb(undefined, typeof options === 'undefined' ? Buffer.from(text, 'utf8') : text);
}

fs.readFile = readFile;
fs.access = access;

module.exports = fs;
