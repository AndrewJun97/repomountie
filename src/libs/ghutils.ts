//
//
// Copyright Â© 2019, 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/common-nodejs-utils';
import yaml from 'js-yaml';
import { Context } from 'probot';
import { COMMIT_FILE_NAMES, REPO_CONFIG_FILE } from '../constants';

interface RepoMountiePullRequestConfig {
  maxLinesChanged: number;
}

interface RepoMountieStaleIssueConfig {
  maxDaysOld: number;
  applyLabel: string;
}

interface RepoCompliance {
  name?: string;
}

export interface RepoMountieConfig {
  pullRequest: RepoMountiePullRequestConfig;
  staleIssue?: RepoMountieStaleIssueConfig;
}

/**
 * Check if a git ref exists
 * Check if a git reference exists; the call to GitHub will fail
 * if the reference does not exists.
 * @param {Context} context The event context context
 * @param {string} ref The ref to be looked up
 * @returns A boolean of true if the ref exists, false otherwise
 */
export const checkIfRefExists = async (context: Context, ref = context.payload.repository.default_branch): Promise<boolean> => {
  try {
    // If the repo does *not* have a master branch then we don't want to add one.
    // The dev team may be doing this off-line and when they go to push master it
    // will cause a conflict because there will be no common root commit.
    await context.github.git.getRef(
      context.repo({
        ref: `heads/${ref}`,
      })
    );

    return true;
  } catch (err) {
    logger.info(`No ref ${ref} exists in ${context.payload.repository.name}`);

    return false;
  }
};

/**
 * Determine if a file exists on a given branch
 * The only way to check if a file exists is to attempt to fetch it;
 * This fn is a wrapper on this capability.
 * @param {Context} context The event context context
 * @param {string} fileName The name of the file to lookup
 * @param {string} ref The name of the branch (Default: master)
 * @returns A true if the file exists, false otherwise.
 */
export const checkIfFileExists = async (context, fileName, ref = 'master'): Promise<boolean> => {
  try {
    await fetchFile(context, fileName, ref);
    return true;
  } catch (err) {
    return false;
  }
};

/**
 * Fetch the contents of a file from GitHub
 * This function will fetch the contents of a file from the latest
 * commit in a ref.
 * @param {Context} context The context of the repo
 * @param {string} fileName The name of the file to be fetched
 * @param {string} ref The ref containing the file (default master)
 * @returns A resolved promise with the `data`, thrown error otherwise.
 */
export const fetchContentsForFile = async (
  context, fileName, ref = context.payload.repository.default_branch
): Promise<any> => {
  try {
    const commits = await context.github.repos.listCommits(
      context.repo({
        path: fileName,
        sha: ref,
      })
    );

    // Not sure if GH returns the results in sorted order.
    // Descending; newest commits are first.
    const lastCommit = commits.data.sort((a, b) => {
      return (new Date(b.commit.committer.date)).getTime() - (new Date(a.commit.committer.date)).getTime();
    }).shift();

    if (!lastCommit) {
      logger.info('Unable to find last commit.');
      return;
    }

    const response = await context.github.repos.getContents(
      context.repo({
        path: fileName,
        ref: lastCommit.sha,
      })
    );

    const data: any = response.data;

    if (data.content && data.type !== 'file') {
      logger.info('Unusable content type retrieved.');
      return;
    }

    return data;
  } catch (err) {
    const message = `Unable to fetch ${fileName}`;
    logger.error(`${message}, error = ${err.message}`);

    throw new Error(message);
  }
};

/**
 * Fetch the repo compliance file
 * The compliance file determines what state any policy compliance
 * is currently in.
 * @param {Context} context The event context context
 * @param {string} fileName The name of the file to fetch
 * @param {string} ref The ref where the file exists
 * @returns A string containing the file data
 */
export const fetchFile = async (
  context, fileName, ref = context.payload.repository.default_branch
): Promise<string> => {
  try {
    const response = await context.github.repos.getContents(
      context.repo({
        branch: ref,
        path: fileName,
      })
    );

    const data: any = response.data;

    if (data.content && data.type !== 'file') {
      throw new Error('No content returned or wrong file type.');
    }

    return Buffer.from(data.content, 'base64').toString();
  } catch (err) {
    const message = `Unable to fetch ${fileName}`;
    logger.error(`${message}, error = ${err.message}`);

    throw new Error(message);
  }
};

/**
 * Fetch the repo compliance file
 * The compliance file determines what state any policy compliance
 * is currently in.
 * @param {Context} context The event context context
 * @returns A `Promise` containing a `RepoCompliance` object
 */
export const fetchComplianceFile = async (context: Context): Promise<RepoCompliance> => {
  try {
    const content = await fetchFile(context, COMMIT_FILE_NAMES.COMPLIANCE);
    return yaml.safeLoad(content);
  } catch (err) {
    const message = 'Unable to fetch compliance file.';
    logger.error(`${message}, error = ${err.message}`);

    throw new Error(message);
  }
};

/**
 * Fetch the repo configuration file
 * The configuration file determines what, if any, cultural policies should
 * be enforced.
 * @param {Context} context The event context context
 * @returns A `Promise` containing a `RepoMountieConfig` object
 */
export const fetchConfigFile = async (context: Context): Promise<RepoMountieConfig> => {
  try {
    const content = await fetchFile(context, REPO_CONFIG_FILE);
    return JSON.parse(content);
  } catch (err) {
    const message = 'Unable to fetch config file.';
    logger.error(`${message}, error = ${err.message}`);

    throw new Error(message);
  }
};

/**
 * Check if a label exists in a given context
 * Check if a given label exists in the repo specified by
 * the `context` argument.
 * @param {Context} context The event context context
 * @param {string} labelName The label name to be checked
 * @returns `true` if the label exists, false otherwise.
 */
export const labelExists = async (
  context: Context, labelName: string
): Promise<boolean> => {
  try {
    const result = await context.github.issues.listLabelsForRepo(context.issue());
    if (!result.data) {
      return false;
    }

    const myMatches = result.data.filter((item) => item.name === labelName);
    return myMatches.length > 0;
  } catch (err) {
    const message = 'Unable to fetch repo labels';
    logger.error(`${message}, error = ${err.message}`);

    return false;
  }
};

/**
 * Add a file to a repo via a pull request
 * Adds a file to a repo via a PR based of the
 * master branch.
 * @param {Context} context The event context context
 * @param {string} commitMessage The commit message for the file
 * @param {string} prTitle The title of the pull request
 * @param {string} prBody The message body of the pull request
 * @param {string} srcBranchName The source branch for the pull request
 * @param {string} fileName The name of the file to be added
 * @param {string} fileData The data of the file to be added
 */
export const addFileViaPullRequest = async (
  context: Context, commitMessage: string, prTitle: string,
  prBody: string, srcBranchName: string, fileName: string,
  fileData: string
) => {
  try {
    // If we don't have a main branch we won't have anywhere
    // to merge the PR.
    const mainbr = await context.github.git.getRef(
      context.repo({
        ref: `heads/${context.payload.repository.default_branch}`,
      })
    );

    // Create a branch to commit to commit the license file
    await context.github.git.createRef(
      context.repo({
        ref: `refs/heads/${srcBranchName}`,
        sha: mainbr.data.object.sha, // where we fork from.
      })
    );

    // Add the file to the new branch
    await context.github.repos.createFile(
      context.repo({
        branch: srcBranchName,
        content: Buffer.from(fileData).toString('base64'),
        message: commitMessage,
        path: fileName,
      })
    );

    // Create a PR to merge the licence ref into master
    await context.github.pulls.create(
      context.repo({
        base: context.payload.repository.default_branch,
        body: prBody,
        head: srcBranchName,
        maintainer_can_modify: true, // maintainers can edit this PR
        title: prTitle,
      })
    );
  } catch (err) {
    const message = `Unable to add ${fileName} file to ${context.payload.repository.name}`;
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

export const fetchCollaborators = async (context): Promise<any[]> => {
  try {
    const results = await context.github.repos.listCollaborators(
      context.repo()
    );

    if (!results && !results.data) {
      return [];
    }

    return results.data;
  } catch (err) {
    const message = `Unable to lookup collaborators in repo ${context.payload.repository.name}`;
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

export const fetchPullRequests = async (
  context, state = 'all'
): Promise<any[]> => {
  try {
    const results = await context.github.pulls.list(
      context.repo({
        state,
      })
    );

    if (!results && !results.data) {
      return [];
    }

    return results.data;
  } catch (err) {
    const message = `Unable to lookup PRs in repo ${context.payload.repository.name}`;
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

/**
 * Check if a pull request exists
 * Check if pull requests exists with a given title
 * @param {Context} context The event context context
 * @param {string} title The title to look for
 * @param {string} state The state the PR must be in.
 * @returns `true` if if a PR exists, false otherwise.
 */
export const hasPullRequestWithTitle = async (
  context, title, state = 'all'
): Promise<boolean> => {
  try {
    const results = await fetchPullRequests(context, state);

    if (results.filter((pr) => pr.title === title).length > 0) {
      return true;
    }

    return false;
  } catch (err) {
    const message = `Unable to lookup PRs in repo ${context.payload.repository.name}`;
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

/**
 * Assign GitHub users to an issue
 * @param context The `Context` containing the GH issue.
 * @param assignees An `Array` of users to assign to the issue
 */
export const assignUsersToIssue = async (
  context: Context, assignees: string[]
) => {
  try {
    await context.github.issues.addAssignees(
      context.issue({
        assignees,
      })
    );
  } catch (err) {
    const message = 'Unable to assign user to issue.';
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

/**
 * Update the contents of a file or create it if non-existent.
 * This fn updates the contents of a file by creating a commit
 * with the appropriate changes.
 * @param {Context} context The event context context
 * @param {string} fileName The name of the file to lookup
 * @returns undefined if successful, throws otherwise
 */
export const updateFileContent = async (
  context: Context, commitMessage: string, srcBranchName: string,
  fileName: string, fileData: string, fileSHA
) => {
  try {
    await context.github.repos.createOrUpdateFile(
      context.repo({
        branch: srcBranchName,
        content: Buffer.from(fileData).toString('base64'),
        message: commitMessage,
        path: fileName,
        sha: fileSHA,
      })
    );
  } catch (err) {
    const message = 'Unable to update file.';
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

/**
 * Check if a user is member of an organization
 * This fn will check if the given user ID belongs to the
 * organization in the given context.
 * @param {Context} context The query context
 * @param {string} userID The GitHub ID of the user
 * @returns True if the user is a member, false otherwise
 */
export const isOrgMember = async (context: Context, userID: string): Promise<boolean> => {
  try {
    const response = await context.github.orgs.checkMembership({
      org: context.payload.organization.login,
      username: userID,
    });

    if (response.status === 204 || response.status === 302) {
      // 204 No Content - The user is a member;
      // 302 Found      - TBD
      return true;
    }

    const message = 'Unexpected return code looking up user';
    logger.info(`${message}, code = ${response.status}`);

    return false;
  } catch (err) {
    // 404 Not Found  - The user is not a member of the org.
    if (err.code === 404) {
      return false;
    }

    const message = 'Unable to lookup user';
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

/**
 * Add a comment to an issue
 * This fn will add a comment to a given issue
 * @param {Context} context The query context
 * @param {string} body The comment body
 * @returns Undefined if successful, thrown error otherwise
 */
export const addCommentToIssue = async (context: Context, body: string) => {
  try {
    await context.github.issues.createComment(
      context.issue({
        body,
      })
    );
  } catch (err) {
    const message = 'Unable to add comment to issue.';
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};

export const blarb = async (
  context: Context, fileName, ref = context.payload.repository.default_branch
) => {
  try {

    // const since = moment(new Date()).subtract(7, 'days').toISOString();
    // Not sure if GH returns the results in sorted order.
    // Descending; newest commits are first.
    const commits = (await context.github.repos.listCommits(
      context.repo({
        path: fileName,
        sha: ref,
        // since,
        // until YYYY-MM-DDTHH:MM:SSZ
      })
    ))
      .data
      .sort((a, b) => {
        return (new Date(b.commit.committer.date)).getTime() - (new Date(a.commit.committer.date)).getTime();
      })
      .slice(-2);

    if (commits.length <= 1) {
      // no change or they probably have not merged the initial the PR yet.
      // This status will be picked up a different way.
      return;
    }

    // a) if the last change is newer than our last known audit check (quarterly); OR
    // b) if the last commit was updated since the last scan (24h ago?).
    // then
    //   if STRA changed, get the previous and current state, as well as the
    //   date of the last commit.
    //   if PIA changed, get the previous and current state, as well as the
    //   date of the last commit.
    //     if the STRA and PIA previous and current state are equal
    //     exit (move along)
    //   send an event with this information.
  } catch (err) {
    const message = 'Unable to add comment to issue.';
    logger.error(`${message}, error = ${err.message}`);

    throw err;
  }
};
