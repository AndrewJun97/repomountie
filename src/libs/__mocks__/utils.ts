//
// SecureImage
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-02-14.
//

import fs from 'fs';
import path from 'path';
import { Context } from 'probot';
import { RepoMountieConfig } from '../utils';

const p0 = path.join(__dirname, '../../../__tests__/fixtures/rmconfig.json');
const config = JSON.parse(fs.readFileSync(p0, 'utf8'));

const p1 = path.join(__dirname, '../../../templates/stale_issue_comment.md');
const template = fs.readFileSync(p1, 'utf8');

export const fetchRepoMountieConfig = async (context: Context): Promise<RepoMountieConfig> => {
  return config;
};

export const loadTemplate = async (path: string): Promise<string> => {
  return template;
}

export const labelExists = async (context: Context, labelName: string): Promise<boolean> => {
  return true;
}