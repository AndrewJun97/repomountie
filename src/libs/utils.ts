//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/common-nodejs-utils';
import fs from 'fs';
import { Context } from 'probot';
import util from 'util';
import { REPO_CONFIG_FILE } from '../constants';

interface RepoMountiePullRequestConfig {
  maxLinesChanged: number;
}

interface RepoMountieStaleIssueConfig {
  maxDaysOld: number;
  applyLabel: string;
}

export interface RepoMountieConfig {
  pullRequest: RepoMountiePullRequestConfig;
  staleIssue?: RepoMountieStaleIssueConfig;
}

export const isJSON = (aString: string): boolean => {
  try {
    JSON.parse(aString);
    return true;
  } catch (err) {
    return false;
  }
}
/**
 * Fetch the repo configuration file
 * The configuration file determines what, if any, cultural policies should
 * be enforced.
 * @param {Context} context The event context context
 * @returns A `RepoMountieConfig` object if one exists
 */
export const fetchRepoMountieConfig = async (context: Context): Promise<RepoMountieConfig> => {
  try {
    const response = await context.github.repos.getContents(
      context.repo({
        branch: 'master',
        path: REPO_CONFIG_FILE,
      })
    );

    const content = Buffer.from(response.data.content, 'base64').toString();
    return JSON.parse(content);
  } catch (err) {
    const message = 'Unable to process config file.';
    logger.error(`${message}, error = ${err.message}`);
    throw new Error(message);
  }
};


/**
 * Load a template file file and return the contents.
 *
 * @param {string} path The path to the template file
 * @returns {Promise<string>} Resolved with contents, rejected otherwise
 */
export const loadTemplate = async (path: string): Promise<string> => {
  const access = util.promisify(fs.access);
  const read = util.promisify(fs.readFile);

  try {
    await access(path, fs.constants.R_OK);
    return read(path, 'utf8');
  } catch (err) {
    const message = `Unable to load template ${path}`;
    throw new Error(`${message}, error = ${err.message}`);
  }
};

/**
 * Extract and return and API response message
 * The errors received from using the github API via Probot
 * will return an Error message with a JSON encoded string in the
 * message property.
 * @param {Error} error The error from which to extract the message
 * @returns Undefined if unparsable, a resolved promise containing the results otherwise
 */
export const extractMessage = async (error: Error): Promise<string> => {
  try {
    if (isJSON(error.message)) {
      const data = JSON.parse(error.message);
      return data.message;
    }

    return error.message;
  } catch (err) {
    const message = 'Unable to extract message from error';
    return message;
  }
};

/**
 * Check if a label exists in a given context
 * Check if a given label exists in the repo specified by
 * the `context` argument.
 * @param {Context} context The event context context
 * @param {string} labelName The label name to be checked
 * @returns `true` if the label exists, false otherwise.
 */
export const labelExists = async (context: Context, labelName: string): Promise<boolean> => {
  try {
    const result = await context.github.issues.listLabelsForRepo(context.issue());
    if (!result.data) {
      return false;
    }

    const myMatches = result.data.filter(item => item.name === labelName);
    return myMatches.length > 0;
  } catch (err) {
    const message = 'Unable to fetch repo labels';
    logger.error(`${message}, error = ${err.message}`);

    return false
  }
}
