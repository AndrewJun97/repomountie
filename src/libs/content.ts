//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/nodejs-common-utils';
import { Context } from 'probot';
import { BRANCHES, TEMPLATES } from '../constants';
import { loadTemplate } from './utils';

/**
 * Add a license file via PR to a given repo
 *
 * @param {Context} context The repo context
 * @returns
 */
export const addLicense = async (context: Context) => {
  const commitMessage: string = 'Add Apache License 2.0';
  const body: string = `
    Repos in our organization need to be licensed under the Apache License 2.0.
    To help you get up to spec I've added one for you as part of this PR;
    please merge it when you can. If you have an exception please add comments to
    this PR and close it without merging it.`;

  try {
    // If we don't have a master we won't have know where to merge the PR
    const master = await context.github.gitdata.getReference(
      context.repo({
        ref: 'heads/master',
      })
    );

    if (!master) {
      return;
    }

    // Create a branch to commit to commit the license file
    await context.github.gitdata.createReference(
      context.repo({
        ref: BRANCHES.LICENSE,
        sha: master.data.object.sha,
      })
    );

    const data = await loadTemplate(TEMPLATES.LICENSE);
    if (!data) {
      return;
    }

    // Add the file to the new branch
    await context.github.repos.createFile(
      context.repo({
        branch: BRANCHES.LICENSE,
        content: Buffer.from(data).toString('base64'),
        message: commitMessage,
        path: 'LICENSE',
      })
    );

    // Create a PR to merge the licence ref into master
    await context.github.pullRequests.create(
      context.repo({
        base: 'master',
        body,
        head: BRANCHES.LICENSE,
        maintainer_can_modify: true, // maintainers cat edit your this PR
        title: 'Add missing license',
      })
    );
  } catch (err) {
    logger.log(err.message);

    throw err;
  }
};
