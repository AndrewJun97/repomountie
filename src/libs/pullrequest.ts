//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-01.
//

import { logger } from '@bcgov/nodejs-common-utils';
import { Context } from 'probot';
import { REPO_CONFIG_FILE, TEXT_FILES } from '../constants';
import { loadTemplate } from '../libs/utils';

interface RepoMountiePullRequestConfig {
  maxLinesChanged: number;
}

interface RepoMountieConfig {
  pullRequest: RepoMountiePullRequestConfig;
}

/**
 * Fetch the repo configuration file
 * The configuration file determines what, if any, cultural policies should
 * be enforced.
 * @param {Context} context The event context context
 * @returns A `RepoMountieConfig` object if one exists
 */
export const fetchRepoMountieConfig = async (context: Context): Promise<RepoMountieConfig> => {
  try {
    const response = await context.github.repos.getContents(
      context.repo({
        branch: 'master',
        path: REPO_CONFIG_FILE,
      })
    );

    const content = Buffer.from(response.data.content, 'base64').toString();
    return JSON.parse(content);
  } catch (err) {
    const message = 'Unable to process config file.';
    logger.error(`${message}, error = ${err.message}`);
    throw new Error(message);
  }
};

/**
 * Validate the length of a pull request
 * The length of a PR is determined by adding the lines deleted and added
 * @param {Context} context The event context context
 * @param {RepoMountieConfig} config The repo config file
 * @returns True if the length is valid, False otherwise
 */
export const isValidPullRequestLength = (context: Context, config: RepoMountieConfig): boolean => {
  const linesChanged =
    context.payload.pull_request.additions + context.payload.pull_request.deletions;

  if (linesChanged <= config.pullRequest.maxLinesChanged) {
    return true;
  }

  return false;
};

/**
 * Validate the PR against codified cultural policies
 * @param {Context} context The event context context
 */
export const validatePullRequestIfRequired = async (context: Context) => {
  try {
    const config = await fetchRepoMountieConfig(context);

    if (!isValidPullRequestLength(context, config)) {
      const rawMessageBody: string = await loadTemplate(TEXT_FILES.HOWTO_PR);
      const messageBody = rawMessageBody
        .replace('[USER_NAME]', context.payload.pull_request.user.login)
        .replace('[MAX_LINES]', `${config.pullRequest.maxLinesChanged}`);

      await context.github.issues.createComment(context.issue({ body: messageBody }));
    }
  } catch (err) {
    const message = 'Unable to validate pull request.';
    logger.error(`${message}, error = ${err.message}`);
  }
};
