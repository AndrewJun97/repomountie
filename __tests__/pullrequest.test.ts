//
// Repo Mountie
//
// Copyright Â© 2018 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2018-10-02.
//

import fs from 'fs';
import path from 'path';
import { Application, Context } from 'probot';
import robot from '../src';
import { extractCommands, isValidPullRequestLength, shouldIgnoredLengthCheck } from '../src/libs/pullrequest';
import { fetchRepoMountieConfig } from '../src/libs/utils';

jest.mock('fs');

const p0 = path.join(__dirname, 'fixtures/pull_request-opened-event.json');
const p1 = path.join(__dirname, 'fixtures/repo-get-content.json');
const p2 = path.join(__dirname, 'fixtures/rmconfig.json');

describe('Repository integration tests', () => {
  let app;
  let github;
  let context;
  let issueOpenedEvent;
  let repoFileContent;
  let repoMountieConfig;

  beforeEach(() => {
    app = new Application();
    app.app = () => 'Token';
    app.load(robot);

    issueOpenedEvent = JSON.parse(fs.readFileSync(p0, 'utf8'));
    repoFileContent = JSON.parse(fs.readFileSync(p1, 'utf8'));
    repoMountieConfig = JSON.parse(fs.readFileSync(p2, 'utf8'));

    github = {
      issues: {
        createComment: jest.fn(),
      },
      repos: {
        createFile: jest.fn(),
        getContents: jest.fn().mockReturnValueOnce(Promise.resolve(repoFileContent)),
      },
    };

    // Passes the mocked out GitHub API into out app instance
    app.auth = () => Promise.resolve(github);

    context = new Context(issueOpenedEvent, github as any, {} as any);
  });

  test('A config file should be fetched from the repo', async () => {
    const response = await fetchRepoMountieConfig(context);

    expect(response).not.toBe(repoMountieConfig);
    expect(github.repos.getContents).toHaveBeenCalled();
  });

  test('A repo with no config should be handled gracefully', async () => {
    const err = new Error('Unable to process config file.');
    const getContents = jest.fn().mockReturnValueOnce(Promise.reject(err));
    github.repos.getContents = getContents;

    await expect(fetchRepoMountieConfig(context)).rejects.toThrow(err);
    expect(github.repos.getContents).toHaveBeenCalled();
  });

  test('A PR with less changes is accepted', () => {
    context.payload.pull_request.additions = 5;
    const myConfig = { ...repoMountieConfig };
    myConfig.pullRequest.maxLinesChanged = 10;

    expect(isValidPullRequestLength(context, myConfig)).toBeTruthy();
  });

  test('A PR with more changes is rejected', () => {
    context.payload.pull_request.additions = 10;
    const myConfig = { ...repoMountieConfig };
    myConfig.pullRequest.maxLinesChanged = 5;

    expect(isValidPullRequestLength(context, myConfig)).toBeFalsy();
  });

  test('A PR with the same number of changes is rejected', () => {
    context.payload.pull_request.additions = 10;
    const myConfig = { ...repoMountieConfig };
    myConfig.pullRequest.maxLinesChanged = 10;

    expect(isValidPullRequestLength(context, myConfig)).toBeTruthy();
  });

  test('A long PR should trigger a help comment to be added', async () => {
    context.payload.pull_request.additions = 1000;
    await app.receive({
      name: 'pull_request.opened',
      payload: issueOpenedEvent.payload,
    });

    expect(github.repos.getContents).toHaveBeenCalled();
    expect(github.issues.createComment).toHaveBeenCalled();
  });

  test('No commands are processed correctly', () => {
    expect(extractCommands(context.payload.pull_request.body).length).toBe(0);
  });

  test('A valid command is extracted', () => {
    context.payload.pull_request.body += '\n /bot-ignore-length';
    expect(extractCommands(context.payload.pull_request.body).length).toBe(1);
  });

  test('Invalid commands are ignored', () => {
    context.payload.pull_request.body += '\n /bot-ignore-length';
    context.payload.pull_request.body += '\n /rm-blarb';

    expect(extractCommands(context.payload.pull_request.body).length).toBe(1);
  });

  test('Empty commands should not be ignored', () => {
    const commands = [];
    expect(shouldIgnoredLengthCheck(commands)).toBeFalsy();
  });

  test('The ignore command should be recognized', () => {
    const commands = ['/bot-ignore-length'];
    expect(shouldIgnoredLengthCheck(commands)).toBeTruthy();
  });

  test('A PR with the ignore command should be ignored', async () => {
    issueOpenedEvent.payload.pull_request.body += '\n /bot-ignore-length';
    await app.receive({
      name: 'pull_request.opened',
      payload: issueOpenedEvent.payload,
    });

    expect(github.repos.getContents).toHaveBeenCalled();
    expect(github.issues.createComment).not.toHaveBeenCalled();
  });
});

// For more information about using TypeScript in your tests, Jest recommends:
// https://github.com/kulshekhar/ts-jest
