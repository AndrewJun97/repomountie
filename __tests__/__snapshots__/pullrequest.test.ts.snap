// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pull requests Repos with stale bot PRs are processed 1`] = `
Array [
  Array [
    Object {
      "body": "Foo 210 bar",
      "issue_number": 2,
      "owner": "bcgov",
      "repo": "hello6",
    },
  ],
]
`;
